File: go/tests/federation_test.go
// federation_test.go
//
// Verifies the CircleAI.Federation port (federation.go): sample-weighted
// averaging + float codecs, the round lifecycle (open/submit/commit) with a
// signature validator, and the HMAC-gated delta dispatcher (accept / duplicate /
// signature-invalid / round-closed).

package circleai_test

import (
	"math"
	"testing"

	circleai "github.com/bhengubv/CircleAI/go"
	"github.com/google/uuid"
)

func TestFederation_WeightedAverage(t *testing.T) {
	// deltas [1,1] (weight 3) and [3,3] (weight 1) -> (3*1 + 1*3)/4 = 1.5 each.
	d1 := circleai.ModelDelta{DeltaPayload: circleai.FederatedEncodeFloats([]float32{1, 1}), SampleCount: 3}
	d2 := circleai.ModelDelta{DeltaPayload: circleai.FederatedEncodeFloats([]float32{3, 3}), SampleCount: 1}
	out, err := circleai.FederatedAverage([]circleai.ModelDelta{d1, d2})
	if err != nil {
		t.Fatalf("average: %v", err)
	}
	got, err := circleai.FederatedDecodeFloats(out)
	if err != nil || len(got) != 2 {
		t.Fatalf("decode = %v err=%v", got, err)
	}
	for _, v := range got {
		if math.Abs(float64(v)-1.5) > 1e-5 {
			t.Fatalf("weighted avg = %v, want 1.5", got)
		}
	}
	if _, err := circleai.FederatedAverage(nil); err == nil {
		t.Fatalf("empty list must error")
	}
}

func TestFederation_RoundLifecycle(t *testing.T) {
	agg := circleai.NewInMemoryFederationAggregator(func(circleai.ModelDelta) bool { return true })
	round, err := agg.OpenRound("m", "1.0", "1.1", 2, 5)
	if err != nil {
		t.Fatalf("open: %v", err)
	}
	mk := func(sc int) circleai.ModelDelta {
		return circleai.ModelDelta{ID: uuid.New(), RoundID: round.ID, DeltaPayload: circleai.FederatedEncodeFloats([]float32{1}), SampleCount: sc}
	}
	// Below MinParticipants -> nil payload.
	if err := agg.SubmitDelta(mk(1)); err != nil {
		t.Fatalf("submit1: %v", err)
	}
	if payload, _ := agg.TryCommit(round.ID); payload != nil {
		t.Fatalf("commit below min must return nil")
	}
	if err := agg.SubmitDelta(mk(2)); err != nil {
		t.Fatalf("submit2: %v", err)
	}
	payload, err := agg.TryCommit(round.ID)
	if err != nil || payload == nil {
		t.Fatalf("commit at min = %v err=%v", payload, err)
	}
	got, _ := agg.GetRound(round.ID)
	if got.Status != circleai.RoundStatusCommitted || got.CurrentParticipantCount != 2 {
		t.Fatalf("round after commit = %+v", got)
	}
	// Idempotent re-commit returns the same payload.
	again, _ := agg.TryCommit(round.ID)
	if len(again) != len(payload) {
		t.Fatalf("re-commit payload differs")
	}
	// Unknown round errors.
	if _, err := agg.GetRound(uuid.New()); err == nil {
		t.Fatalf("unknown round must error")
	}
}

func TestFederation_DeltaDispatcher(t *testing.T) {
	key := []byte("0123456789abcdef0123456789abcdef")
	agg := circleai.NewInMemoryFederationAggregator(circleai.HMACSignatureValidator(key))
	round, _ := agg.OpenRound("m", "1.0", "1.1", 1, 5)
	disp := circleai.NewInMemoryFederationDeltaDispatcher(agg, circleai.HMACSignatureValidator(key))

	good := circleai.ModelDelta{ID: uuid.New(), RoundID: round.ID, ContributorUhid: "u", ModelID: "m", FromVersion: "1.0", DeltaPayload: circleai.FederatedEncodeFloats([]float32{1})}
	good.Signature = circleai.HMACSignDelta(key, good)

	if outcome, _ := disp.VerifyAndSubmit(good); outcome != circleai.DeltaAccepted {
		t.Fatalf("first submit outcome = %d, want Accepted", outcome)
	}
	if outcome, _ := disp.VerifyAndSubmit(good); outcome != circleai.DeltaDuplicate {
		t.Fatalf("replay outcome = %d, want Duplicate", outcome)
	}
	// Tampered signature.
	bad := good
	bad.ID = uuid.New()
	bad.Signature = []byte("wrong")
	if outcome, _ := disp.VerifyAndSubmit(bad); outcome != circleai.DeltaSignatureInvalid {
		t.Fatalf("bad-sig outcome = %d, want SignatureInvalid", outcome)
	}
	// Unknown round.
	stray := good
	stray.ID = uuid.New()
	stray.RoundID = uuid.New()
	stray.Signature = circleai.HMACSignDelta(key, stray)
	if outcome, _ := disp.VerifyAndSubmit(stray); outcome != circleai.DeltaRoundUnknown {
		t.Fatalf("unknown-round outcome = %d, want RoundUnknown", outcome)
	}
}

File: typescript/tests/federation.test.ts
// federation.test.ts
// Verifies the CircleAI.Federation port: little-endian float encode/decode round
// trips, sample-size-weighted averaging (with Math.fround at the float write
// site), and the InMemoryFederationAggregator round lifecycle (open → submit →
// commit with min-participants gating and signature filtering).

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  FederatedAveraging,
  InMemoryFederationAggregator,
  RoundStatus,
  DeltaDispatchOutcome,
  modelDelta,
  type ModelDelta,
} from "../src/federation/index";

function delta(roundId: string, samples: number, floats: number[], id = crypto.randomUUID()): ModelDelta {
  return modelDelta(
    id,
    roundId,
    "uhid",
    "m",
    "1.0.0",
    FederatedAveraging.encodeFloats(floats),
    samples,
    new Uint8Array([1, 2, 3]),
    new Date(),
  );
}

describe("FederatedAveraging encode/decode", () => {
  it("round-trips float arrays through Math.fround", () => {
    const values = [1.5, -2.25, 0, 3.14159];
    const decoded = FederatedAveraging.decodeFloats(FederatedAveraging.encodeFloats(values));
    assert.equal(decoded.length, 4);
    assert.equal(decoded[0], 1.5);
    assert.equal(decoded[1], -2.25);
    assert.equal(decoded[3], Math.fround(3.14159));
  });

  it("rejects a payload whose length is not a multiple of 4", () => {
    assert.throws(() => FederatedAveraging.decodeFloats(new Uint8Array([1, 2, 3])));
  });
});

describe("FederatedAveraging.average", () => {
  it("computes a sample-size-weighted average", () => {
    // weights 1:3 over [0,0] and [4,8] → [3,6].
    const d1 = delta("r", 1, [0, 0]);
    const d2 = delta("r", 3, [4, 8]);
    const out = FederatedAveraging.decodeFloats(FederatedAveraging.average([d1, d2]));
    assert.equal(out[0], 3);
    assert.equal(out[1], 6);
  });

  it("throws on empty / mismatched / zero-weight inputs", () => {
    assert.throws(() => FederatedAveraging.average([]));
    assert.throws(() => FederatedAveraging.average([delta("r", 1, [1, 2]), delta("r", 1, [1])]));
    assert.throws(() => FederatedAveraging.average([delta("r", 0, [1, 2])]));
  });
});

describe("InMemoryFederationAggregator", () => {
  it("opens a round with validated parameters", async () => {
    const agg = new InMemoryFederationAggregator(() => true);
    const round = await agg.openRoundAsync("m", "1.0.0", "1.1.0", 2, 5);
    assert.equal(round.status, RoundStatus.Open);
    assert.equal(round.minParticipants, 2);
    await assert.rejects(() => agg.openRoundAsync("m", "1", "2", 0, 5));
    await assert.rejects(() => agg.openRoundAsync("m", "1", "2", 3, 2));
  });

  it("commits once min participants are met and averages valid deltas", async () => {
    const agg = new InMemoryFederationAggregator(() => true);
    const round = await agg.openRoundAsync("m", "1.0.0", "1.1.0", 2, 5);
    await agg.submitDeltaAsync(delta(round.id, 1, [0, 0]));
    // Below min → null.
    assert.equal(await agg.tryCommitAsync(round.id), null);
    await agg.submitDeltaAsync(delta(round.id, 3, [4, 8]));
    const committed = await agg.tryCommitAsync(round.id);
    assert.ok(committed);
    const out = FederatedAveraging.decodeFloats(committed as Uint8Array);
    assert.deepEqual(out, [3, 6]);
    assert.equal((await agg.getRoundAsync(round.id)).status, RoundStatus.Committed);
  });

  it("re-returns the committed payload idempotently", async () => {
    const agg = new InMemoryFederationAggregator(() => true);
    const round = await agg.openRoundAsync("m", "1", "2", 1, 5);
    await agg.submitDeltaAsync(delta(round.id, 1, [2, 2]));
    const first = await agg.tryCommitAsync(round.id);
    const second = await agg.tryCommitAsync(round.id);
    assert.deepEqual([...(first as Uint8Array)], [...(second as Uint8Array)]);
  });

  it("drops signature-invalid deltas at commit time", async () => {
    // Only the delta with samples===2 validates; min is 1, so it commits with just that.
    const agg = new InMemoryFederationAggregator((d) => d.sampleCount === 2);
    const round = await agg.openRoundAsync("m", "1", "2", 1, 5);
    await agg.submitDeltaAsync(delta(round.id, 1, [9, 9]));
    await agg.submitDeltaAsync(delta(round.id, 2, [4, 4]));
    const out = FederatedAveraging.decodeFloats((await agg.tryCommitAsync(round.id)) as Uint8Array);
    assert.deepEqual(out, [4, 4]);
  });

  it("rejects submissions to an unknown round and past MaxParticipants", async () => {
    const agg = new InMemoryFederationAggregator(() => true);
    await assert.rejects(() => agg.submitDeltaAsync(delta("nope", 1, [1, 1])));
    const round = await agg.openRoundAsync("m", "1", "2", 1, 1);
    await agg.submitDeltaAsync(delta(round.id, 1, [1, 1]));
    await assert.rejects(() => agg.submitDeltaAsync(delta(round.id, 1, [2, 2])));
  });

  it("ignores empty-payload deltas without counting them", async () => {
    const agg = new InMemoryFederationAggregator(() => true);
    const round = await agg.openRoundAsync("m", "1", "2", 1, 5);
    await agg.submitDeltaAsync(
      modelDelta(crypto.randomUUID(), round.id, "u", "m", "1", new Uint8Array(0), 5, new Uint8Array(), new Date()),
    );
    assert.equal((await agg.getRoundAsync(round.id)).currentParticipantCount, 0);
  });

  it("exposes DeltaDispatchOutcome enum values", () => {
    assert.equal(DeltaDispatchOutcome.Accepted, 0);
    assert.equal(DeltaDispatchOutcome.Duplicate, 2);
  });
});
